### PR TITLE
add drag and drop

### PR DESCRIPTION
**Pull request by SeremTitus**
todo:

- [x] add in between drop
- [x]  add drop location indicator

### CODE REVIEW

**Comment by SeremTitus** [comment]
-  extend on #218

**Comment by SeremTitus** [comment]
- [ ] fix hover number field while draging changes number field value

**Comment by SeremTitus** [comment]
> * extend on [Implement move_tags_to in TagSVG #218](https://github.com/MewPurPur/GodSVG/pull/218)

- [ ] waiting for full working implimentation of  #218

**Comment by MewPurPur** [comment]
What's the status on this? I haven't tested it yet because it's draft.

**Comment by SeremTitus** [comment]
Ready For Review

**Comment by MewPurPur** [comment]
I'll test this tomorrow, in the meantime, could you adjust your code to the [GDScript style guide](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html) and the few additional guidelines in [CONTRIBUTING.md](https://github.com/MewPurPur/GodSVG/blob/main/CONTRIBUTING.md)?

**Comment by MewPurPur** [comment]
This is quite a complex feature and I'm impressed by how it looks so far! However, a few fixes are needed before I can accept it.

If you hover just below the first tag or just above the last tag while dragging them, the drop position appears in the wrong place.

image

Operations like pressing Delete or Ctrl+Down while you're dragging a tag cause weird behaviors or crashes, they need to be blocked somehow.

Dragging a tag into the viewport causes the viewport to be dragged around and the mouse cursor doesn't go in MOUSE_CURSOR_FORBIDDEN mode.

image

The indicator causes the tag it's under to get a worse-looking stylebox, I don't think that should be a thing. The depth tint also disappears. You could look into using draw_stylebox() to draw an indicator separate from the stylebox of the tag itself.

image

Inside Indications.svg, you need to have proper logic for changing the selected TIDs after the tags are moved.

LineEdits and widgets like that shouldn't block the dropping.

image


-----

Also, could you please rebase again and do the codestyle changes as mentioned above?

**Comment by SeremTitus** [comment]
Thanks for the detailed review, I have undertaken some refactoring:
# Refactors
1. Tags are selected on click
> ... If you hover just below the first tag or just above the last tag while dragging them, the drop position appears in the wrong place...
2. Fix drop position error 
3. Operations like pressing Delete or Ctrl + Down  are now blocked while dragging
> ... The indicator causes the tag it's under to get a worse-looking stylebox...
4.  StyleBox maintain there style, achieved by turning root node into a PanelContainer with now provides the drop indicator style
> ... Dragging a tag into the viewport causes the viewport to be dragged around...
5.  The display viewport dragging is locked  when tag is draged
6. I have gone through both style guides and tried to structure the code as was defined, kindly highlight areas where it does not fall with in the style guide if there is and I will improve on it
7. Also rebased

# Limitations
> ... Dragging a tag into the viewport ... mouse cursor doesn't go in MOUSE_CURSOR_FORBIDDEN mode...
1. Godot limitation the Cursor does not change by itself to FORBIDDEN mode, I tried to change it in code using set_default_cursor_shape(value) method in Control class did not work (might be a Godot bug not sure)
> ... LineEdits and widgets like that shouldn't block the dropping...
2.  Currently there is no mode ( at least that I know off) that a child Control Node can be place into in order to pass down drop request to its parent (Godot limitation)


> ... Inside Indications.gd, you need to have proper logic for changing the selected TIDs after the tags are moved...
- Kindly Elaborate, It currently handed that all moved tags are selected, others that where previously selected are no longer selected if not moved... Could you describe the desired behavour?

**Comment by MewPurPur** [comment]
Selections are not retained most of the time after you finish dragging.

There are still codestyle issues with whitespace and comments should use sentence case.

**Comment by MewPurPur** [comment]
> Tags are now selected when you release the mouse, instead of when you press. This should be reverted as it feels unresponsive.

Still not fixed.

Tags now jitter a bit because the node structure itself changes now. But that's problematic:

https://github.com/MewPurPur/GodSVG/assets/85438892/c16b5d51-9b18-41f3-ac4f-74d1580713c9

The correct solution here is to find a way to use draw_stylebox() to draw a yellow stylebox on top of the tag, instead of putting it inside and causing it to resize.

-----

I recently changed LineEdits to pass the mouse instead of stop, so that when hovered or focused, the tag remains focused. This is a mixed bag here:

- Now you can drop tags on top of them without it showing as forbidden, good!

- When you drag them without a selection (i.e. to make a selection) it will drag the whole tag.

I think something that would work is to go in BetterLineEdit, detect the gui inputs of drag and drop, and do `accept_event()` so these specific ones don't propagate.

-----

The selection adjustments, a lot of work is needed here. For example, dragging a tag inside the one below it will unselect it. Now there's some invalid TID inside Indications.gd variables, which makes it easy to get into invalid states (pressing Delete crashes GodSVG for example, because you're trying to delete inexistent tags)

I suggest playtesting this really hard inside Indications.gd `_on_tags_moved_to` by checking variables like `selected_tids` before and after.

**Comment by MewPurPur** [comment]
If you need help with any of the above, lmk.

**Comment by SeremTitus** [comment]
> Tags are now selected when you release the mouse, instead of when you press. This should be reverted as it feels unresponsive

^ I might need help there, I this might need a rework on how selection works at least on click

if setting to LineEdits to pass the mouse instead of stop, allows to drop on top of it ... this can be set  to when drag begins and aand back to original state when end (_notification function in tagEditor can do that for all input fields), I will work on that and the jitters and playtesting Indications.gd further

**Comment by SeremTitus** [comment]
new
Fix

**Comment by SeremTitus** [comment]
https://github.com/MewPurPur/GodSVG/assets/58259212/b108ca96-d0ce-4f51-b6b3-9adec796b95c

Fix jitter

**Comment by MewPurPur** [comment]
It seems like you're struggling to understand what to do for `Indications._on_tags_moved_to()`. If you want, I can merge this now and iron out these kinks myself.

**Comment by SeremTitus** [comment]
> It seems like you're struggling to understand what to do for `Indications._on_tags_moved_to()`. If you want, I can merge this now and iron out these kinks myself?

I agree with you on that

File: src/ui_parts/TagsContainer.gd [hunk 1]
```diff
@@ -0,0 +1,65 @@
+extends VBoxContainer
+
+const safe_margin: float = 0.18 # 0 - 1
```
Comment (MewPurPur) [inline src/ui_parts/TagsContainer.gd:3]:
```suggestion
const safe_margin = 0.18
```

File: src/ui_parts/TagsContainer.gd [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+@onready var scroll_container: ScrollContainer = %ScrollContainer
+
+var is_drag_begin:bool = false
```
Comment (MewPurPur) [inline src/ui_parts/TagsContainer.gd:7]:
```suggestion
var is_drag_begin := false
```

File: src/ui_parts/TagsContainer.gd [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+
+func _process(_delta: float) -> void:
+	# Scroll with moving draged object
```
Comment (MewPurPur) [inline src/ui_parts/TagsContainer.gd:11]:
```suggestion
	# Scroll with moving draged object.
```

File: src/ui_parts/TagsContainer.gd [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	if  not scroll_container == null and is_drag_begin:
```
Comment (MewPurPur) [inline src/ui_parts/TagsContainer.gd:12]:
```suggestion
	if scroll_container != null and is_drag_begin:
```

File: src/ui_parts/TagsContainer.gd [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		var working_area := scroll_container.get_global_rect()
+		var shrink_ratio := safe_margin * float(working_area.size.y)
+		var safe_area := working_area.grow_individual(0,- shrink_ratio,
+											0,- shrink_ratio)
+		working_area = working_area.grow_individual(0,shrink_ratio/3,
+									0,shrink_ratio/3)
```
Comment (MewPurPur) [inline src/ui_parts/TagsContainer.gd:18]:
```suggestion
		var safe_area := working_area.grow_individual(0, -shrink_ratio, 0, -shrink_ratio)
		working_area = working_area.grow_individual(0, shrink_ratio/3, 0, shrink_ratio/3)
```

File: src/ui_parts/TagsContainer.gd [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+		var mouse_position := get_global_mouse_position()
+		if ( working_area.has_point(mouse_position) and
+			not safe_area.has_point(mouse_position)
+		):
```
Comment (MewPurPur) [inline src/ui_parts/TagsContainer.gd:22]:
```suggestion
		var mouse_pos := get_global_mouse_position()
		if working_area.has_point(mouse_pos) and not safe_area.has_point(mouse_pos):
```

File: src/ui_parts/TagsContainer.gd [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+			if safe_area.position.y < mouse_position.y:
+				scroll_container.scroll_vertical += 5
+			else:
+				scroll_container.scroll_vertical -= 5
+
+
+func _can_drop_data(_at_position: Vector2, current_tid: Variant) -> bool:
+	if current_tid is Array[PackedInt32Array]:
+		var child := drop_location_calculator()
+		if child:
```
Comment (MewPurPur) [inline src/ui_parts/TagsContainer.gd:32]:
```suggestion
		if child != null:
```

File: src/ui_parts/TagsContainer.gd [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+
+func _notification(what: int) -> void:
+	if what == NOTIFICATION_DRAG_BEGIN:
+		is_drag_begin = true
+	elif what == NOTIFICATION_DRAG_END:
+		is_drag_begin = false
+
+
+func drop_location_calculator() -> Control:
```
Comment (MewPurPur) [inline src/ui_parts/TagsContainer.gd:58]: Let's rename this to `calculate_drop_location()`.

File: src/ui_parts/TagsContainer.gd [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+	var at_position: Vector2 = get_global_mouse_position()
```
Comment (MewPurPur) [inline src/ui_parts/TagsContainer.gd:59]:
```suggestion
	var at_position := get_global_mouse_position()
```

File: src/ui_parts/tag_editor.gd [hunk 10]
```diff
@@ -14,19 +14,26 @@ const PathField = preload("res://src/ui_elements/path_field.tscn")
...
 
 var tid: PackedInt32Array
 var tag: Tag
 
+enum DropState{
+	Inside = 0,
+	Up,
+	Down,
+	Outside,
+}
```
Comment (MewPurPur) [inline src/ui_parts/tag_editor.gd:35]:
```suggestion
enum DropState {INSIDE, UP, DOWN, OUTSIDE}
```

File: src/ui_parts/tag_editor.gd [hunk 11]
```diff
@@ -87,9 +94,61 @@ func _ready() -> void:
...
 
 
+func _get_drag_data(_at_position: Vector2) -> Variant:
+	var data: Array[PackedInt32Array] = [tid]
+	if tid in Indications.selected_tids:
+		data = Indications.selected_tids
+	data = Utils.filter_tids_descendant(data)
+	var tags_container: VBoxContainer = VBoxContainer.new()
+	for drag_tid in data:
+		var preview = TagEditor.instantiate()
```
Comment (MewPurPur) [inline src/ui_parts/tag_editor.gd:104]:
```suggestion
		var preview := TagEditor.instantiate()
```

File: src/ui_parts/tag_editor.gd [hunk 12]
```diff
@@ -87,9 +94,61 @@ func _ready() -> void:
...
+	var state := drop_location_calculator(get_global_mouse_position())
+	var new_tid := tid.duplicate()
+	match state:
+		DropState.Inside:
+			new_tid.append(0)
+		DropState.Up:
+			new_tid[-1] -= 1 if new_tid[-1] > 0 else 0
+		DropState.Down:
+			new_tid[-1] += 1
+	SVG.root_tag.move_tags_to(current_tid,new_tid)
```
Comment (MewPurPur) [inline src/ui_parts/tag_editor.gd:137]:
```suggestion
	SVG.root_tag.move_tags_to(current_tid, new_tid)
```

**Review by MewPurPur — COMMENTED** [review]
Some code style remarks. I didn't have the time to finish this, but please look over to see if there are more things that don't abide by GDScript's style guide and the few additional ones in CONTRIBUTORS.md

**Comment by MewPurPur** [comment]
You've only committed the suggested changes without the adjacent things, which isn't a good practice. When I made suggestions like using CAPITAL_CASE for enums, of course I also implied that you should update the enum values everywhere they are used.

**Comment by SeremTitus** [comment]
> You've only committed the suggested changes without testing, which isn't a good practice

My bad, my intention was to do that then to pull with all changes for testing...
Retested fully working , code refactored with suggested changes and rebased

**Comment by MewPurPur** [comment]
Is this good for a - hopefully - last round of testing?

**Comment by SeremTitus** [comment]
Yes

**Comment by MewPurPur** [comment]
I got a crash in paths, should be an easy fix though.

> Tags are now selected when you release the mouse, instead of when you press. This should be reverted as it feels unresponsive.

Still not fixed.

**Comment by SeremTitus** [comment]
Select  on click is a difficult  one but I will give it an attempt , I may have to change how the selection  system  works

**Comment by MewPurPur** [comment]
OK, after that is solved, I'll merge this and fix the other issues.

**Comment by SeremTitus** [comment]
> OK, after that is solved, I'll merge this and fix the other issues.

Select on click is fix, inspired from Godot Editor's FileSystem ( if there is no item selected selection is on click, if there item selected selection is on release  and drag from any item drags only previously selected) our approach currently is slightly adaptation of that such as when there is selected items clicked item is added to the selected if drag is immediately initialized else just select the item on click...this is tested and ready for review testing

My opinion on Selection on drop , is that once dropped they should  be all selected (as implemented currently) For a case such that you have dropped tags and what to drag them again it be faster if you do not have to select them again

**Comment by SeremTitus** [comment]
> I got a consistent crash when using paths
This let me work on that bug

**Comment by MewPurPur** [comment]
Thanks!

**Pull request merged**